text/0000-placement-by-return.md

```diff
@@ -0,0 +1,1097 @@
+
+Guaranteed Copy Elision only kicks in when a GCE-applicable expression is directly returned from a function, either by being the function's last expression or by being the expression part of a `return` statement.
+
+## The unsafe method of writing return values
+
+The `std::mem` module exposes two symmetrical intrinsics that can allow you to achieve Copy Elision in any circumstance.
+
+### `write_return_with`
+
+    pub unsafe fn write_return_with<T, F: for<'a> FnOnce(*mut u8)>(f: F) -> T {
+        write_unsized_return_with(Layout::new::<T>(), |p| {f(p); &mut *(p as *mut T)})
+    }
+    extern "intrinsic" pub unsafe fn write_unsized_return_with<T: ?Sized, F: for<'a> FnOnce(*mut u8) -> &mut T>(layout: Layout, f: F) -> T;
+
+[write_return_with]: #write_return_with
+
+Directly access the caller-provided return slot.
+
+The return slot is an uninitialized memory space provided by a function's caller to write the return value, and implicitly passed as a pointer. A function which directly returns the result of evaluating another function, like `fn foo() -> ReturnedValue { bar() }`, may simply pass the slot along, thus avoiding expensive copies. In cases where it is impossible to implement a function by directly returning the value of a function call or a primitive expression, directly accessing the return slot and writing it directly may be necessary.
+
+Return slots are not always used; small values like pointers, numbers, and booleans may be returned by filling a register instead. In this case, `write_return_with` will implicitly create a "return slot", pass a pointer to it to `f`, and then copy the return value from the "return slot" to the register.
+
+The pointer returned by `write_unsized_return_with`'s callback should be the same as the one provided to it, along any necessary unsize information (such as the slice's length or the trait object's vtable pointer). Additionally, when using `write_unsized_return_with` for sized types, the provided layout must be exactly the same as the one produced by `Layout::new<T>()`. These two values must be correct even when they're redundant. A function that returns a slice may return a slice that is smaller than the requested allocation Layout, in case it is unable to predict the amount of data that will be available to it, but when producing a trait object, it must know exactly the right size for its allocation.
+
+#### Panicking
+
+`f` is allowed to panic. If it panics, the underlying memory should still be freed (the built-in collections are well-behaved), but the return slot implementation will assume that the allocated memory was not initialized, *and shall not call T's destructor*. If `f` allocates additional resources itself before panicking, it is responsible for freeing it.
```

Comment:
This seems weird-- don't panics always leak memory unless they're caught?

Reply:
No, in most cases, panics shouldn't leak memory. Panicking still runs the drop glue for everything on the stack being unwound through.

Reply:
Ah, that makes sense-- I completely forgot about drop glue.